src/modules/languages/rust.nix

```diff
@@ -13,67 +10,91 @@ let
-          };
-        }
-        // genAttrs tools (name: lib.mkOption {
-          type = lib.types.package;
-          default = pkgs.${name};
-          defaultText = lib.literalExpression "pkgs.${name}";
-          description = "${name} package";
-        });
+        freeformType = lib.types.attrsOf lib.types.package;
+
+        options =
+          let
+            documented-components = [ "rustc" "cargo" "clippy" "rustfmt" "rust-analyzer" ];
+            mkComponentOption = component: lib.mkOption {
+              type = lib.types.nullOr lib.types.package;
+              default = pkgs.${component};
+              defaultText = lib.literalExpression "pkgs.${component}";
+              description = "${component} package";
+            };
+          in
+          lib.genAttrs documented-components mkComponentOption;
       });
-      defaultText = lib.literalExpression "pkgs";
       default = { };
-      description = "Attribute set of packages including rustc and Cargo.";
-    };
-
-    version = lib.mkOption {
-      type = lib.types.nullOr lib.types.str;
-      default = null;
-      description = "Set to stable, beta, or latest.";
+      defaultText = lib.literalExpression "nixpkgs";
+      description = "Rust component packages. May optionally define additional components, for example `miri`.";
     };
   };
 
   config = lib.mkMerge [
     (lib.mkIf cfg.enable {
-      packages = attrValues (getAttrs tools cfg.packages) ++ lib.optional pkgs.stdenv.isDarwin pkgs.libiconv;
+      packages = (builtins.map (c: cfg.toolchain.${c} or (throw (error "toolchain.${c}"))) cfg.components)
+        ++ lib.optional pkgs.stdenv.isDarwin pkgs.libiconv;
 
       # enable compiler tooling by default to expose things like cc
       languages.c.enable = lib.mkDefault true;
 
-      env.RUST_SRC_PATH = cfg.packages.rust-src;
+      # RUST_SRC_PATH is necessary when rust-src is not at the same location as
+      # as rustc. This is the case with the rust toolchain from nixpkgs.
+      env.RUST_SRC_PATH =
+        if cfg.toolchain ? rust-src
+        then "${cfg.toolchain.rust-src}/lib/rustlib/src/rust/library"
+        else pkgs.rustPlatform.rustLibSrc;
 
-      pre-commit.tools.cargo = lib.mkForce cfg.packages.cargo;
-      pre-commit.tools.rustfmt = lib.mkForce cfg.packages.rustfmt;
-      pre-commit.tools.clippy = lib.mkForce cfg.packages.clippy;
+      pre-commit.tools.cargo = cfg.toolchain.cargo or null;
+      pre-commit.tools.rustfmt = cfg.toolchain.rustfmt or null;
+      pre-commit.tools.clippy = cfg.toolchain.clippy or null;
     })
     (lib.mkIf (cfg.enable && pkgs.stdenv.isDarwin) {
       env.RUSTFLAGS = [ "-L framework=${config.env.DEVENV_PROFILE}/Library/Frameworks" ];
       env.RUSTDOCFLAGS = [ "-L framework=${config.env.DEVENV_PROFILE}/Library/Frameworks" ];
       env.CFLAGS = [ "-iframework ${config.env.DEVENV_PROFILE}/Library/Frameworks" ];
     })
-    (lib.mkIf (cfg.version != null) (
+    (lib.mkIf (cfg.channel != null) (
```

Comment:
`cfg.channel != "nixpkgs"`

Reply:
ACK, haven't tested the latest changes yet 😅